File: android/Reguerta/app/src/main/java/com/reguerta/user/data/notifications/FirestoreNotificationRepository.kt
package com.reguerta.user.data.notifications

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.access.MemberRole
import com.reguerta.user.domain.notifications.NotificationEvent
import com.reguerta.user.domain.notifications.NotificationContentPolicy
import com.reguerta.user.domain.notifications.NotificationRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

class FirestoreNotificationRepository(
    private val firestore: FirebaseFirestore,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : NotificationRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val notificationsCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.NOTIFICATION_EVENTS)

    private fun notificationInboxCollectionPath(memberId: String): String =
        "${firestorePath.documentPath(ReguertaFirestoreCollection.USERS, memberId)}/notificationInbox"

    private fun notificationReadsCollectionPath(memberId: String): String =
        "${firestorePath.documentPath(ReguertaFirestoreCollection.USERS, memberId)}/notificationReads"

    override suspend fun getNotificationsFor(member: Member): List<NotificationEvent> = withContext(Dispatchers.IO) {
        try {
            val snapshot = firestore.collection(notificationInboxCollectionPath(member.id))
                .get()
                .await()
            decodeNotificationDocuments(
                documents = snapshot.documents.map { document ->
                    document.id to document.data.orEmpty()
                },
                source = NotificationDocumentSource.INBOX,
            )
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "notificationInbox")
        }
    }

    override suspend fun getReadNotificationIds(memberId: String): Set<String> = withContext(Dispatchers.IO) {
        try {
            val snapshot = firestore.collection(notificationReadsCollectionPath(memberId)).get().await()
            snapshot.documents.map { it.id }.toSet()
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "notificationReads")
        }
    }

    override suspend fun markNotificationsRead(
        memberId: String,
        notificationIds: Set<String>,
        readAtMillis: Long,
    ) = withContext(Dispatchers.IO) {
        val normalizedIds = notificationIds.map(String::trim).filter(String::isNotBlank).toSet()
        if (normalizedIds.isEmpty()) return@withContext

        val batch = firestore.batch()
        val readAt = Timestamp(
            readAtMillis / 1_000,
            ((readAtMillis % 1_000) * 1_000_000).toInt(),
        )
        val collection = firestore.collection(notificationReadsCollectionPath(memberId))
        normalizedIds.forEach { notificationId ->
            batch.set(
                collection.document(notificationId),
                mapOf(
                    "notificationEventId" to notificationId,
                    "readAt" to readAt,
                ),
                SetOptions.merge(),
            )
        }
        try {
            batch.commit().await()
            Unit
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "notificationReads")
        }
    }

    override suspend fun sendNotification(event: NotificationEvent): NotificationEvent = withContext(Dispatchers.IO) {
        val documentId = event.id.ifBlank {
            firestore.collection(notificationsCollectionPath).document().id
        }
        val persisted = event.copy(id = documentId)
        val payload = mutableMapOf<String, Any?>(
            "title" to persisted.title,
            "body" to persisted.body,
            "type" to persisted.type,
            "target" to persisted.target,
            "sentAt" to Timestamp(
                persisted.sentAtMillis / 1_000,
                ((persisted.sentAtMillis % 1_000) * 1_000_000).toInt(),
            ),
            "createdBy" to persisted.createdBy,
        )
        persisted.weekKey?.let { payload["weekKey"] = it }
        val targetPayload = when (persisted.target) {
            "users" -> mapOf("userIds" to persisted.userIds)
            "segment" -> when (persisted.segmentType) {
                "role" -> mapOf(
                    "segmentType" to "role",
                    "role" to persisted.targetRole?.wireValue(),
                )
                else -> emptyMap<String, Any?>()
            }
            else -> emptyMap<String, Any?>()
        }
        payload["targetPayload"] = targetPayload

        try {
            firestore.collection(notificationsCollectionPath)
                .document(documentId)
                .set(payload, SetOptions.merge())
                .await()
            persisted
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "notificationEvents/$documentId")
        }
    }
}

internal enum class NotificationDocumentSource(val resourceCollection: String) {
    EVENT("notificationEvents"),
    INBOX("notificationInbox"),
}

internal fun decodeNotificationDocuments(
    documents: List<Pair<String, Map<String, Any?>>>,
    source: NotificationDocumentSource,
): List<NotificationEvent> = documents
    .map { (documentId, data) -> decodeNotificationDocument(documentId, data, source) }
    .sortedByDescending(NotificationEvent::sentAtMillis)

internal fun decodeNotificationDocument(
    documentId: String,
    data: Map<String, Any?>,
    source: NotificationDocumentSource,
): NotificationEvent = NotificationDocumentDto.decode(documentId, data, source).toDomain()

internal data class NotificationDocumentDto(
    val id: String,
    val title: String,
    val body: String,
    val type: String,
    val target: String,
    val userIds: List<String>,
    val segmentType: String?,
    val targetRole: MemberRole?,
    val createdBy: String,
    val sentAtMillis: Long,
    val weekKey: String?,
    val contentPolicy: NotificationContentPolicy,
) {
    fun toDomain(): NotificationEvent = NotificationEvent(
        id = id,
        title = title,
        body = body,
        type = type,
        target = target,
        userIds = userIds,
        segmentType = segmentType,
        targetRole = targetRole,
        createdBy = createdBy,
        sentAtMillis = sentAtMillis,
        weekKey = weekKey,
        contentPolicy = contentPolicy,
    )

    companion object {
        fun decode(
            documentId: String,
            data: Map<String, Any?>,
            source: NotificationDocumentSource,
        ): NotificationDocumentDto {
            if (documentId.isBlank() || documentId != documentId.trim()) {
                invalidNotificationDocument(source, documentId)
            }
            if (source == NotificationDocumentSource.INBOX &&
                data.requiredExactNotificationString("notificationEventId", source, documentId) != documentId
            ) {
                invalidNotificationDocument(source, documentId)
            }
            val type = data.requiredExactNotificationString("type", source, documentId)
            if (type !in CANONICAL_NOTIFICATION_TYPES) {
                invalidNotificationDocument(source, documentId)
            }
            val target = data.requiredExactNotificationString("target", source, documentId)
            val audience = data.requiredNotificationAudience(target, source, documentId)
            val title = data.requiredNotificationString("title", source, documentId)
            val body = data.requiredNotificationString("body", source, documentId)
            val createdBy = data.requiredNotificationString("createdBy", source, documentId)
            val weekKey = data.optionalNotificationString("weekKey", source, documentId)
            return NotificationDocumentDto(
                id = documentId,
                title = title,
                body = body,
                type = type,
                target = target,
                userIds = audience.userIds,
                segmentType = audience.segmentType,
                targetRole = audience.targetRole,
                createdBy = createdBy,
                sentAtMillis = data.requiredNotificationTimestampMillis("sentAt", source, documentId),
                weekKey = weekKey,
                contentPolicy = data.notificationContentPolicy(
                    title = title,
                    body = body,
                    type = type,
                    target = target,
                    audience = audience,
                    createdBy = createdBy,
                    weekKey = weekKey,
                    source = source,
                    documentId = documentId,
                ),
            )
        }
    }
}

private data class DecodedNotificationAudience(
    val userIds: List<String> = emptyList(),
    val segmentType: String? = null,
    val targetRole: MemberRole? = null,
)

private fun Map<String, Any?>.notificationContentPolicy(
    title: String,
    body: String,
    type: String,
    target: String,
    audience: DecodedNotificationAudience,
    createdBy: String,
    weekKey: String?,
    source: NotificationDocumentSource,
    documentId: String,
): NotificationContentPolicy {
    val fields = setOf("schemaVersion", "operationKind", "contentPolicy")
    val present = keys.intersect(fields)
    if (present.isEmpty()) return NotificationContentPolicy.EMBEDDED
    val schemaVersion = this["schemaVersion"] as? Number
    if (
        present != fields ||
        schemaVersion?.toDouble() != 1.0 ||
        schemaVersion.toLong() != 1L ||
        this["operationKind"] != "shiftPlanningNotification" ||
        this["contentPolicy"] != "genericReferenceOnly" ||
        title != "Turnos actualizados" ||
        body != "Consulta la aplicación para ver la información actualizada." ||
        type != "shift_updated" ||
        target != "users" ||
        audience.userIds.size != 1 ||
        audience.segmentType != null ||
        audience.targetRole != null ||
        createdBy != "system" ||
        weekKey != null
    ) {
        invalidNotificationDocument(source, documentId)
    }
    return NotificationContentPolicy.AUTHORIZED_FETCH_REQUIRED
}

private fun Map<String, Any?>.requiredNotificationAudience(
    target: String,
    source: NotificationDocumentSource,
    documentId: String,
): DecodedNotificationAudience {
    val payload = this["targetPayload"] as? Map<*, *>
        ?: invalidNotificationDocument(source, documentId)
    return when (target) {
        "all" -> {
            if (payload.isNotEmpty()) invalidNotificationDocument(source, documentId)
            DecodedNotificationAudience()
        }
        "users" -> {
            if (payload.keys != setOf("userIds")) invalidNotificationDocument(source, documentId)
            val rawUserIds = payload["userIds"] as? List<*>
                ?: invalidNotificationDocument(source, documentId)
            val userIds = rawUserIds.map { value ->
                (value as? String)?.trim()?.ifEmpty {
                    invalidNotificationDocument(source, documentId)
                } ?: invalidNotificationDocument(source, documentId)
            }
            if (userIds.isEmpty()) invalidNotificationDocument(source, documentId)
            DecodedNotificationAudience(userIds = userIds)
        }
        "segment" -> {
            if (payload.keys != setOf("segmentType", "role")) {
                invalidNotificationDocument(source, documentId)
            }
            if (payload["segmentType"] != "role") {
                invalidNotificationDocument(source, documentId)
            }
            val role = when (payload["role"]) {
                "member" -> MemberRole.MEMBER
                "producer" -> MemberRole.PRODUCER
                "admin" -> MemberRole.ADMIN
                else -> invalidNotificationDocument(source, documentId)
            }
            DecodedNotificationAudience(segmentType = "role", targetRole = role)
        }
        else -> invalidNotificationDocument(source, documentId)
    }
}

private fun Map<String, Any?>.requiredNotificationString(
    field: String,
    source: NotificationDocumentSource,
    documentId: String,
): String = optionalNotificationString(field, source, documentId)
    ?: invalidNotificationDocument(source, documentId)

private fun Map<String, Any?>.requiredExactNotificationString(
    field: String,
    source: NotificationDocumentSource,
    documentId: String,
): String {
    val value = this[field] as? String ?: invalidNotificationDocument(source, documentId)
    if (value.isEmpty() || value != value.trim()) invalidNotificationDocument(source, documentId)
    return value
}

private fun Map<String, Any?>.optionalNotificationString(
    field: String,
    source: NotificationDocumentSource,
    documentId: String,
): String? {
    val value = this[field] ?: return null
    if (value !is String) invalidNotificationDocument(source, documentId)
    return value.trim().ifEmpty { invalidNotificationDocument(source, documentId) }
}

private fun Map<String, Any?>.requiredNotificationTimestampMillis(
    field: String,
    source: NotificationDocumentSource,
    documentId: String,
): Long = (this[field] as? Timestamp)?.toDate()?.time
    ?: invalidNotificationDocument(source, documentId)

private fun invalidNotificationDocument(
    source: NotificationDocumentSource,
    documentId: String,
): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "${source.resourceCollection}/$documentId",
)

private val CANONICAL_NOTIFICATION_TYPES = setOf(
    "order_reminder",
    "order_auto_generated",
    "shift_swap_requested",
    "shift_swap_available",
    "shift_swap_unavailable",
    "shift_swap_accepted",
    "shift_swap_applied",
    "shift_updated",
    "news_published",
    "admin_broadcast",
)

private fun MemberRole.wireValue(): String =
    when (this) {
        MemberRole.MEMBER -> "member"
        MemberRole.PRODUCER -> "producer"
        MemberRole.ADMIN -> "admin"
    }

File: ios/Reguerta/Reguerta/Data/Notifications/FirestoreNotificationRepository.swift
import FirebaseFirestore
import Foundation

enum FirestoreNotificationDocumentSource {
    case events
    case inbox

    var collectionName: String {
        switch self {
        case .events:
            "notificationEvents"
        case .inbox:
            "notificationInbox"
        }
    }
}

final class FirestoreNotificationRepository: @unchecked Sendable, NotificationRepository {
    private let db: Firestore
    private let environment: ReguertaFirestoreEnvironment?

    init(db: Firestore = Firestore.firestore(), environment: ReguertaFirestoreEnvironment? = nil) {
        self.db = db
        self.environment = environment
    }

    private var notificationsCollection: CollectionReference {
        db.reguertaCollection(.notificationEvents, environment: environment)
    }

    private func notificationReadsCollection(memberId: String) -> CollectionReference {
        db.reguertaCollection(.users, environment: environment)
            .document(memberId)
            .collection("notificationReads")
    }

    private func notificationInboxCollection(memberId: String) -> CollectionReference {
        db.reguertaCollection(.users, environment: environment)
            .document(memberId)
            .collection("notificationInbox")
    }

    func notifications(visibleTo member: Member) async throws -> [NotificationEvent] {
        do {
            let snapshot = try await notificationInboxCollection(memberId: member.id).getDocuments()
            return try Self.notificationEvents(
                documents: snapshot.documents.map {
                    (documentID: $0.documentID, data: $0.data())
                },
                source: .inbox
            )
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "notificationInbox")
        }
    }

    func allNotifications() async throws -> [NotificationEvent] {
        do {
            let snapshot = try await notificationsCollection.getDocuments()
            return try Self.notificationEvents(
                documents: snapshot.documents.map {
                    (documentID: $0.documentID, data: $0.data())
                },
                source: .events
            )
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "notificationEvents")
        }
    }

    func readNotificationIds(memberId: String) async throws -> Set<String> {
        do {
            let snapshot = try await notificationReadsCollection(memberId: memberId).getDocuments()
            return Set(snapshot.documents.map(\.documentID))
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "notificationReads")
        }
    }

    func markNotificationsRead(memberId: String, notificationIds: [String], readAtMillis: Int64) async throws {
        let normalizedIds = Set(notificationIds.map {
            $0.trimmingCharacters(in: .whitespacesAndNewlines)
        }.filter { !$0.isEmpty })
        guard !normalizedIds.isEmpty else { return }

        let batch = db.batch()
        let readAt = Timestamp(date: Date(timeIntervalSince1970: TimeInterval(readAtMillis) / 1_000))
        let collection = notificationReadsCollection(memberId: memberId)
        for notificationId in normalizedIds {
            batch.setData(
                [
                    "notificationEventId": notificationId,
                    "readAt": readAt
                ],
                forDocument: collection.document(notificationId),
                merge: true
            )
        }
        do {
            try await batch.commit()
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "notificationReads")
        }
    }

    func send(event: NotificationEvent) async throws -> NotificationEvent {
        let documentId = event.id.isEmpty ? notificationsCollection.document().documentID : event.id
        let persisted = NotificationEvent(
            id: documentId,
            title: event.title,
            body: event.body,
            type: event.type,
            target: event.target,
            userIds: event.userIds,
            segmentType: event.segmentType,
            targetRole: event.targetRole,
            createdBy: event.createdBy,
            sentAtMillis: event.sentAtMillis,
            weekKey: event.weekKey
        )

        let targetPayload: [String: Any?]
        switch persisted.target {
        case "users":
            targetPayload = ["userIds": persisted.userIds]
        case "segment" where persisted.segmentType == "role":
            targetPayload = [
                "segmentType": "role",
                "role": persisted.targetRole?.wireValue
            ]
        default:
            targetPayload = [:]
        }

        var payload: [String: Any] = [
            "title": persisted.title,
            "body": persisted.body,
            "type": persisted.type,
            "target": persisted.target,
            "targetPayload": targetPayload,
            "sentAt": Timestamp(date: Date(timeIntervalSince1970: TimeInterval(persisted.sentAtMillis) / 1_000)),
            "createdBy": persisted.createdBy
        ]
        if let weekKey = persisted.weekKey {
            payload["weekKey"] = weekKey
        }
        do {
            try await notificationsCollection.document(documentId).setData(payload, merge: true)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(
                error,
                resource: "notificationEvents/\(documentId)"
            )
        }
        return persisted
    }

    static func notificationEvents(
        documents: [(documentID: String, data: [String: Any])],
        source: FirestoreNotificationDocumentSource
    ) throws -> [NotificationEvent] {
        try documents
            .map {
                try notificationEvent(
                    documentID: $0.documentID,
                    data: $0.data,
                    source: source
                )
            }
            .sorted { $0.sentAtMillis > $1.sentAtMillis }
    }

    static func notificationEvent(
        documentID: String,
        data: [String: Any],
        source: FirestoreNotificationDocumentSource
    ) throws -> NotificationEvent {
        let dto = try FirestoreNotificationDocumentDecoder.decode(
            documentID: documentID,
            data: data,
            source: source
        )
        return FirestoreNotificationDocumentMapper.toDomain(dto)
    }
}

private extension MemberRole {
    var wireValue: String {
        switch self {
        case .member: "member"
        case .producer: "producer"
        case .admin: "admin"
        }
    }
}

private struct FirestoreNotificationDocumentDTO {
    let documentID: String
    let title: String
    let body: String
    let type: String
    let target: String
    let userIDs: [String]
    let segmentType: String?
    let targetRole: MemberRole?
    let createdBy: String
    let sentAtMillis: Int64
    let weekKey: String?
}

private struct FirestoreNotificationAudienceDTO {
    let userIDs: [String]
    let segmentType: String?
    let role: MemberRole?
}

private enum FirestoreNotificationDocumentDecoder {
    private static let canonicalTypes: Set<String> = [
        "order_reminder",
        "order_auto_generated",
        "shift_swap_requested",
        "shift_swap_available",
        "shift_swap_unavailable",
        "shift_swap_accepted",
        "shift_swap_applied",
        "shift_updated",
        "news_published",
        "admin_broadcast"
    ]

    static func decode(
        documentID: String,
        data: [String: Any],
        source: FirestoreNotificationDocumentSource
    ) throws -> FirestoreNotificationDocumentDTO {
        let normalizedDocumentID = documentID.trimmingCharacters(in: .whitespacesAndNewlines)
        let resource = "\(source.collectionName)/\(documentID)"
        guard !documentID.isEmpty, documentID == normalizedDocumentID else {
            throw RepositoryError.invalidData(resource: resource)
        }
        if source == .inbox {
            let eventID = try exactRequiredString(
                data,
                field: "notificationEventId",
                resource: resource
            )
            guard eventID == documentID else {
                throw RepositoryError.invalidData(resource: resource)
            }
        }

        let type = try exactRequiredString(data, field: "type", resource: resource)
        guard canonicalTypes.contains(type) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        let target = try exactRequiredString(data, field: "target", resource: resource)
        let audience = try decodeAudience(data, target: target, resource: resource)

        return try FirestoreNotificationDocumentDTO(
            documentID: documentID,
            title: requiredString(data, field: "title", resource: resource),
            body: requiredString(data, field: "body", resource: resource),
            type: type,
            target: target,
            userIDs: audience.userIDs,
            segmentType: audience.segmentType,
            targetRole: audience.role,
            createdBy: requiredString(data, field: "createdBy", resource: resource),
            sentAtMillis: requiredTimestampMillis(data, field: "sentAt", resource: resource),
            weekKey: optionalString(data, field: "weekKey", resource: resource)
        )
    }

    private static func decodeAudience(
        _ data: [String: Any],
        target: String,
        resource: String
    ) throws -> FirestoreNotificationAudienceDTO {
        guard let payload = data["targetPayload"] as? [String: Any] else {
            throw RepositoryError.invalidData(resource: resource)
        }
        switch target {
        case "all":
            guard payload.isEmpty else {
                throw RepositoryError.invalidData(resource: resource)
            }
            return FirestoreNotificationAudienceDTO(
                userIDs: [],
                segmentType: nil,
                role: nil
            )
        case "users":
            guard Set(payload.keys) == ["userIds"],
                  let rawUserIDs = payload["userIds"] as? [String],
                  !rawUserIDs.isEmpty else {
                throw RepositoryError.invalidData(resource: resource)
            }
            let userIDs = rawUserIDs.map {
                $0.trimmingCharacters(in: .whitespacesAndNewlines)
            }
            guard userIDs.allSatisfy({ !$0.isEmpty }) else {
                throw RepositoryError.invalidData(resource: resource)
            }
            return FirestoreNotificationAudienceDTO(
                userIDs: userIDs,
                segmentType: nil,
                role: nil
            )
        case "segment":
            guard Set(payload.keys) == ["segmentType", "role"],
                  let segmentType = payload["segmentType"] as? String,
                  segmentType == "role",
                  let roleValue = payload["role"] as? String,
                  let role = MemberRole(rawValue: roleValue) else {
                throw RepositoryError.invalidData(resource: resource)
            }
            return FirestoreNotificationAudienceDTO(
                userIDs: [],
                segmentType: segmentType,
                role: role
            )
        default:
            throw RepositoryError.invalidData(resource: resource)
        }
    }

    private static func exactRequiredString(_ data: [String: Any], field: String, resource: String) throws -> String {
        guard let string = data[field] as? String, !string.isEmpty else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return string
    }

    private static func requiredString(_ data: [String: Any], field: String, resource: String) throws -> String {
        guard let value = try optionalString(data, field: field, resource: resource) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return value
    }

    private static func optionalString(_ data: [String: Any], field: String, resource: String) throws -> String? {
        guard let value = data[field] else { return nil }
        if value is NSNull { return nil }
        guard let string = value as? String else {
            throw RepositoryError.invalidData(resource: resource)
        }
        let normalized = string.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return normalized
    }

    private static func requiredTimestampMillis(
        _ data: [String: Any],
        field: String,
        resource: String
    ) throws -> Int64 {
        guard let timestamp = data[field] as? Timestamp else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return Int64(timestamp.dateValue().timeIntervalSince1970 * 1_000)
    }
}

private enum FirestoreNotificationDocumentMapper {
    static func toDomain(_ dto: FirestoreNotificationDocumentDTO) -> NotificationEvent {
        NotificationEvent(
            id: dto.documentID,
            title: dto.title,
            body: dto.body,
            type: dto.type,
            target: dto.target,
            userIds: dto.userIDs,
            segmentType: dto.segmentType,
            targetRole: dto.targetRole,
            createdBy: dto.createdBy,
            sentAtMillis: dto.sentAtMillis,
            weekKey: dto.weekKey
        )
    }
}
